### PR TITLE
fix(sns-subscriptions): encrypted SQS subscribed to SNS don't receive data

### LINKED ISSUE #11122: [cloudwatch] SQS connected to SNS don't receive data when using KMS
When SQS is configured with QueueEncryption.KMS and master key is set, it does not successfully receive data from SNS that it is subscribed to.

### Reproduction Steps

```python
topic = aws_sns.Topic(self.stack, 'alarm-topic',
                              display_name='Topic for alarms',
                              topic_name=self.alarms_topic_name)
queue = aws_sqs.Queue(self.stack, 'alarm-queue',
                              encryption=aws_sqs.QueueEncryption.KMS,
                              encryption_master_key=key,
                              retention_period=core.Duration.days(7),
                              queue_name=self.alarms_queue_name
                              )
topic.add_subscription(aws_sns_subscriptions.SqsSubscription(queue))
```

### What did you expect to happen?

Successfully retrieve message from SQS

### What actually happened?

When the SNS has delivery logs enabled, the following failure appears in the log:

> {
    "notification": {
        "messageMD5Sum": "2bc17714a33dbaf9dbe6a57a8410eb7f",
        "messageId": "054099fb-9d8e-5f40-b657-c1eb69e60b29",
        "topicArn": "[SNS ARN]",
        "timestamp": "2020-10-26 15:58:32.47"
    },
    "delivery": {
        "deliveryId": "c894a538-64e0-5577-b85d-fd2890888698",
        "destination": "[SQS ARN],
        "providerResponse": "{\"ErrorCode\":\"KMS.AccessDeniedException\",\"ErrorMessage\":\"null (Service: AWSKMS; Status Code: 400; Error Code: AccessDeniedException; Request ID: 9582607f-3c84-4f8e-8d4b-fa9932c21146; Proxy: null)\",\"sqsRequestId\":\"Unrecoverable\"}",
        "dwellTimeMs": 66,
        "attempts": 1,
        "statusCode": 400
    },
    "status": "FAILURE"
}

### Environment

CLI Version : 1.70
Framework Version: Python 3.8.5
Node.js Version: v14.11.0
OS : Linux
Language (Version): Python (3.8.5)

### Other

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by jumi-dev**
Grant KMS permission to SNS service principal if the queue is encrypted.

fixes #11122

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts [hunk 1]
```diff
@@ -46,6 +46,11 @@ export class SqsSubscription implements sns.ITopicSubscription {
       },
     }));
 
+    // grant KMS permission to SNS service principal if the queue is encrypted
+    if (this.queue.encryptionMasterKey) {
+      this.queue.encryptionMasterKey.grantEncryptDecrypt(new iam.ServicePrincipal('sns.amazonaws.com'));
```
Comment (MrArnoldPalmer) [inline packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts:51]: I'm pretty sure there is a way to limit this using the arn of the sns topic in a conditional so we can limit the scope of resources being granted permissions to the key. I'm still looking around for this but let me know if this is something you've tried already. 

Additionally my instinct is that we would only need "encrypt" here, though all the documentation seems to point to needing the decrypt permissions as well.

**Review by MrArnoldPalmer — CHANGES_REQUESTED** [review]
Just dependency change for now. Looking into scoping the permissions principal as well.

File: packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts (hunk 1 above)
Comment (jumi-dev) [reply packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts:51]: > I'm pretty sure there is a way to limit this using the arn of the sns topic in a conditional so we can limit the scope of resources being granted permissions to the key. I'm still looking around for this but let me know if this is something you've tried already.

I didn't find a solution for it yet. The SQS Queue policy supports property `aws:SourceArn` to specify a SNS ARN (see 
https://docs.aws.amazon.com/sns/latest/dg/sns-access-policy-use-cases.html).

The same attribute isn't available in KMS:
_AWS KMS supports all AWS global condition keys except for the following ones: aws:SourceArn_
(see https://docs.aws.amazon.com/kms/latest/developerguide/policy-conditions.html)

I will also try to find another way to restrict the permission.

> Additionally my instinct is that we would only need "encrypt" here, though all the documentation seems to point to needing the decrypt permissions as well.

Yes, I tried to use only "encrypt" which wasn't sufficient. Therefore, the documentation seems to be correct.
Is it fine to use `grantEncryptDecrypt`? Or is it better to grant only the actions `kms:Decrypt` and `kms:GenerateDataKey*` from the documentation?

File: packages/@aws-cdk/aws-sns-subscriptions/package.json [hunk 2]
```diff
@@ -73,6 +73,7 @@
   },
   "dependencies": {
     "@aws-cdk/aws-iam": "0.0.0",
+    "@aws-cdk/aws-kms": "0.0.0",
```
[the remark this answers is not in the document]
Comment (jumi-dev) [reply packages/@aws-cdk/aws-sns-subscriptions/package.json]: Yes, you're right. I've fixed it.

File: packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts (hunk 1 above)
Comment (MrArnoldPalmer) [reply packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts:51]: `grantEncrypDecrypt` is fine if that is what we need which it seems like we do. I'll ask around a bit more about scoping to further and get back to you in the next couple days.. Let me know if you find a solution in the meantime.

**Comment by TheRealAmazonKendra** [comment]
This change was merged as part of PR #14960. Closing this.

File: packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts (hunk 1 above)
Comment (HansFalkenberg-Visma) [reply packages/@aws-cdk/aws-sns-subscriptions/lib/sqs.ts:51]: SourceArn can now be used to restrict KMS access the same way as SQS access. 

https://docs.aws.amazon.com/kms/latest/developerguide/policy-conditions.html
> To help prevent an AWS service from being used as a confused deputy in a policy where the principal is an AWS service principal, you can use the aws:SourceArn or aws:SourceAccount  global condition keys.

I added the following to my KMS policy
```
            "Condition": {
                "ArnEquals": {
                    "aws:SourceArn": "arn:aws:sns:eu-west-1:ACCOUNT:TOPIC"
                }
            }
```

When it matches the source topic it works, if it doesn't match the source topic, the SNS fails with this error:
```
"providerResponse": "{\"ErrorCode\":\"KMS.AccessDeniedException\",\"ErrorMessage\":\"null (Service: AWSKMS; Status Code: 400; Error Code: AccessDeniedException; Request ID: 40edd241-7474-4b44-98bf-de5ba09013fb; Proxy: null)\",\"sqsRequestId\":\"Unrecoverable\"}"
```

**Pull request closed without merging**